// shrimpy-node/dist/dto-converters/backtest-asset-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_dto_converter_1 = require("./date-dto-converter");
var BacktestAssetDtoConverter = /** @class */ (function () {
    function BacktestAssetDtoConverter() {
        this._dateDtoConverter = new date_dto_converter_1.DateDtoConverter();
    }
    BacktestAssetDtoConverter.prototype.convertFromDto = function (dto) {
        var result = {
            symbol: dto.symbol,
            endTime: this._dateDtoConverter.convertFromDto(dto.endTime),
            startTime: this._dateDtoConverter.convertFromDto(dto.startTime),
        };
        return result;
    };
    BacktestAssetDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            symbol: model.symbol,
            endTime: this._dateDtoConverter.convertToDto(model.endTime),
            startTime: this._dateDtoConverter.convertToDto(model.startTime),
        };
        return result;
    };
    return BacktestAssetDtoConverter;
}());
exports.BacktestAssetDtoConverter = BacktestAssetDtoConverter;
