// shrimpy-node/dist/dto-converters/balance-change-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var decimal_dto_converter_1 = require("./decimal-dto-converter");
var BalanceChangeDtoConverter = /** @class */ (function () {
    function BalanceChangeDtoConverter() {
        this._decimalDtoConverter = new decimal_dto_converter_1.DecimalDtoConverter();
    }
    BalanceChangeDtoConverter.prototype.convertFromDto = function (dto) {
        var result = {
            btcValue: dto.btcValue,
            nativeValue: this._decimalDtoConverter.convertFromDto(dto.nativeValue),
            symbol: dto.symbol,
            usdValue: dto.usdValue,
        };
        return result;
    };
    BalanceChangeDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            btcValue: model.btcValue,
            nativeValue: this._decimalDtoConverter.convertToDto(model.nativeValue),
            symbol: model.symbol,
            usdValue: model.usdValue,
        };
        return result;
    };
    return BalanceChangeDtoConverter;
}());
exports.BalanceChangeDtoConverter = BalanceChangeDtoConverter;
