// shrimpy-node/dist/dto-converters/dynamic-strategy-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var decimal_dto_converter_1 = require("./decimal-dto-converter");
var DynamicStrategyDtoConverter = /** @class */ (function () {
    function DynamicStrategyDtoConverter() {
        this._decimalDtoConveter = new decimal_dto_converter_1.DecimalDtoConverter();
    }
    DynamicStrategyDtoConverter.prototype.convertFromDto = function (dto) {
        var result = {
            excludedSymbols: dto.excludedSymbols.slice(0),
            isDynamic: dto.isDynamic,
            isEqualWeight: dto.isEqualWeight,
            maxPercent: this._decimalDtoConveter.convertFromDto(dto.maxPercent),
            minPercent: this._decimalDtoConveter.convertFromDto(dto.minPercent),
            topAssetCount: dto.topAssetCount,
        };
        return result;
    };
    DynamicStrategyDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            excludedSymbols: model.excludedSymbols.slice(0),
            isDynamic: model.isDynamic,
            isEqualWeight: model.isEqualWeight,
            maxPercent: this._decimalDtoConveter.convertToDto(model.maxPercent),
            minPercent: this._decimalDtoConveter.convertToDto(model.minPercent),
            topAssetCount: model.topAssetCount,
        };
        return result;
    };
    return DynamicStrategyDtoConverter;
}());
exports.DynamicStrategyDtoConverter = DynamicStrategyDtoConverter;
