// shrimpy-node/dist/dto-converters/historical-instruments-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_dto_converter_1 = require("./date-dto-converter");
var HistoricalInstrumentsDtoConverter = /** @class */ (function () {
    function HistoricalInstrumentsDtoConverter() {
        this._dateDtoConverter = new date_dto_converter_1.DateDtoConverter();
    }
    HistoricalInstrumentsDtoConverter.prototype.convertFromDto = function (dto) {
        var _this = this;
        return dto.map(function (dtoItem) {
            var orderBookStartTime = dtoItem.orderBookStartTime !== null ? _this._dateDtoConverter.convertFromDto(dtoItem.orderBookStartTime) : null;
            var orderBookEndTime = dtoItem.orderBookEndTime !== null ? _this._dateDtoConverter.convertFromDto(dtoItem.orderBookEndTime) : null;
            var tradeStartTime = dtoItem.tradeStartTime !== null ? _this._dateDtoConverter.convertFromDto(dtoItem.tradeStartTime) : null;
            var tradeEndTime = dtoItem.tradeEndTime !== null ? _this._dateDtoConverter.convertFromDto(dtoItem.tradeEndTime) : null;
            var result = {
                exchange: dtoItem.exchange,
                baseTradingSymbol: dtoItem.baseTradingSymbol,
                quoteTradingSymbol: dtoItem.quoteTradingSymbol,
                orderBookStartTime: orderBookStartTime,
                orderBookEndTime: orderBookEndTime,
                tradeStartTime: tradeStartTime,
                tradeEndTime: tradeEndTime
            };
            return result;
        });
    };
    HistoricalInstrumentsDtoConverter.prototype.convertToDto = function (model) {
        var _this = this;
        return model.map(function (model) {
            var orderBookStartTime = model.orderBookStartTime !== null ? _this._dateDtoConverter.convertToDto(model.orderBookStartTime) : null;
            var orderBookEndTime = model.orderBookEndTime !== null ? _this._dateDtoConverter.convertToDto(model.orderBookEndTime) : null;
            var tradeStartTime = model.tradeStartTime !== null ? _this._dateDtoConverter.convertToDto(model.tradeStartTime) : null;
            var tradeEndTime = model.tradeEndTime !== null ? _this._dateDtoConverter.convertToDto(model.tradeEndTime) : null;
            var result = {
                exchange: model.exchange,
                baseTradingSymbol: model.baseTradingSymbol,
                quoteTradingSymbol: model.quoteTradingSymbol,
                orderBookStartTime: orderBookStartTime,
                orderBookEndTime: orderBookEndTime,
                tradeStartTime: tradeStartTime,
                tradeEndTime: tradeEndTime
            };
            return result;
        });
    };
    return HistoricalInstrumentsDtoConverter;
}());
exports.HistoricalInstrumentsDtoConverter = HistoricalInstrumentsDtoConverter;
