// shrimpy-node/dist/dto-converters/historical-candlestick-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_dto_converter_1 = require("./date-dto-converter");
var decimal_dto_converter_1 = require("./decimal-dto-converter");
var HistoricalCandlestickDtoConverter = /** @class */ (function () {
    function HistoricalCandlestickDtoConverter() {
        this._dateDtoConverter = new date_dto_converter_1.DateDtoConverter();
        this._decimalDtoConverter = new decimal_dto_converter_1.DecimalDtoConverter();
    }
    HistoricalCandlestickDtoConverter.prototype.convertFromDto = function (dto) {
        var result = {
            open: this._decimalDtoConverter.convertFromDto(dto.open),
            high: this._decimalDtoConverter.convertFromDto(dto.high),
            low: this._decimalDtoConverter.convertFromDto(dto.low),
            close: this._decimalDtoConverter.convertFromDto(dto.close),
            volume: this._decimalDtoConverter.convertFromDto(dto.volume),
            quoteVolume: dto.quoteVolume,
            btcVolume: dto.btcVolume,
            usdVolume: dto.usdVolume,
            time: this._dateDtoConverter.convertFromDto(dto.time),
        };
        return result;
    };
    HistoricalCandlestickDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            open: this._decimalDtoConverter.convertToDto(model.open),
            high: this._decimalDtoConverter.convertToDto(model.high),
            low: this._decimalDtoConverter.convertToDto(model.low),
            close: this._decimalDtoConverter.convertToDto(model.close),
            volume: this._decimalDtoConverter.convertToDto(model.volume),
            quoteVolume: model.quoteVolume,
            btcVolume: model.btcVolume,
            usdVolume: model.usdVolume,
            time: this._dateDtoConverter.convertToDto(model.time),
        };
        return result;
    };
    return HistoricalCandlestickDtoConverter;
}());
exports.HistoricalCandlestickDtoConverter = HistoricalCandlestickDtoConverter;
