// shrimpy-node/dist/dto-converters/market-order-books-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var exchange_order_book_dto_converter_1 = require("./exchange-order-book-dto-converter");
var MarketOrderBooksDtoConverter = /** @class */ (function () {
    function MarketOrderBooksDtoConverter() {
        this._exchangeOrderBookDtoConverter = new exchange_order_book_dto_converter_1.ExchangeOrderBookDtoConverter();
    }
    MarketOrderBooksDtoConverter.prototype.convertFromDto = function (dto) {
        var _this = this;
        var orderBooks = dto.orderBooks.map(function (exchangeOrderBookDto) {
            return _this._exchangeOrderBookDtoConverter.convertFromDto(exchangeOrderBookDto);
        });
        var result = {
            baseSymbol: dto.baseSymbol,
            orderBooks: orderBooks,
            quoteSymbol: dto.quoteSymbol
        };
        return result;
    };
    MarketOrderBooksDtoConverter.prototype.convertToDto = function (model) {
        var _this = this;
        var orderBookDtos = model.orderBooks.map(function (exchangeOrderBook) {
            return _this._exchangeOrderBookDtoConverter.convertToDto(exchangeOrderBook);
        });
        var result = {
            baseSymbol: model.baseSymbol,
            orderBooks: orderBookDtos,
            quoteSymbol: model.quoteSymbol
        };
        return result;
    };
    return MarketOrderBooksDtoConverter;
}());
exports.MarketOrderBooksDtoConverter = MarketOrderBooksDtoConverter;
