// shrimpy-node/dist/dto-converters/predictions-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var prediction_item_dto_converter_1 = require("./prediction-item-dto-converter");
var PredictionsDtoConverter = /** @class */ (function () {
    function PredictionsDtoConverter() {
        this._predictionItemDtoConverter = new prediction_item_dto_converter_1.PredictionItemDtoConverter();
    }
    PredictionsDtoConverter.prototype.convertFromDto = function (dto) {
        var _this = this;
        var predictions = dto.predictions.map(function (predictionsDto) {
            return _this._predictionItemDtoConverter.convertFromDto(predictionsDto);
        });
        var result = {
            predictions: predictions,
            baseTradingSymbol: dto.baseTradingSymbol,
            quoteTradingSymbol: dto.quoteTradingSymbol,
        };
        return result;
    };
    PredictionsDtoConverter.prototype.convertToDto = function (model) {
        var _this = this;
        var predictions = model.predictions.map(function (predictions) {
            return _this._predictionItemDtoConverter.convertToDto(predictions);
        });
        var result = {
            baseTradingSymbol: model.baseTradingSymbol,
            quoteTradingSymbol: model.quoteTradingSymbol,
            predictions: predictions
        };
        return result;
    };
    return PredictionsDtoConverter;
}());
exports.PredictionsDtoConverter = PredictionsDtoConverter;
