// shrimpy-node/dist/dto-converters/total-balance-history-item-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_dto_converter_1 = require("./date-dto-converter");
var TotalBalanceHistoryItemDtoConverter = /** @class */ (function () {
    function TotalBalanceHistoryItemDtoConverter() {
        this._dateDtoConverter = new date_dto_converter_1.DateDtoConverter();
    }
    TotalBalanceHistoryItemDtoConverter.prototype.convertFromDto = function (dto) {
        var result = {
            btcValue: dto.btcValue,
            date: this._dateDtoConverter.convertFromDto(dto.date),
            usdValue: dto.usdValue,
        };
        return result;
    };
    TotalBalanceHistoryItemDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            btcValue: model.btcValue,
            date: this._dateDtoConverter.convertToDto(model.date),
            usdValue: model.usdValue,
        };
        return result;
    };
    return TotalBalanceHistoryItemDtoConverter;
}());
exports.TotalBalanceHistoryItemDtoConverter = TotalBalanceHistoryItemDtoConverter;
