// shrimpy-node/dist/dto-converters/trade-fill-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var decimal_dto_converter_1 = require("./decimal-dto-converter");
var TradeFillDtoConverter = /** @class */ (function () {
    function TradeFillDtoConverter() {
        this._decimalDtoConverter = new decimal_dto_converter_1.DecimalDtoConverter();
    }
    TradeFillDtoConverter.prototype.convertFromDto = function (dto) {
        var result = {
            baseAmount: this._decimalDtoConverter.convertFromDto(dto.baseAmount),
            baseSymbol: dto.baseSymbol,
            btcValue: dto.btcValue,
            price: this._decimalDtoConverter.convertFromDto(dto.price),
            quoteAmount: this._decimalDtoConverter.convertFromDto(dto.quoteAmount),
            quoteSymbol: dto.quoteSymbol,
            side: dto.side,
            usdValue: dto.usdValue,
        };
        return result;
    };
    TradeFillDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            baseAmount: this._decimalDtoConverter.convertToDto(model.baseAmount),
            baseSymbol: model.baseSymbol,
            btcValue: model.btcValue,
            price: this._decimalDtoConverter.convertToDto(model.price),
            quoteAmount: this._decimalDtoConverter.convertToDto(model.quoteAmount),
            quoteSymbol: model.quoteSymbol,
            side: model.side,
            usdValue: model.usdValue,
        };
        return result;
    };
    return TradeFillDtoConverter;
}());
exports.TradeFillDtoConverter = TradeFillDtoConverter;
