// shrimpy-node/dist/dto-converters/trade-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var decimal_dto_converter_1 = require("./decimal-dto-converter");
var TradeDtoConverter = /** @class */ (function () {
    function TradeDtoConverter() {
        this._decimalDtoConverter = new decimal_dto_converter_1.DecimalDtoConverter();
    }
    TradeDtoConverter.prototype.convertFromDto = function (dto) {
        var result = {
            amount: this._decimalDtoConverter.convertFromDto(dto.amount),
            errorCode: dto.errorCode,
            errorMessage: dto.errorMessage,
            fromSymbol: dto.fromSymbol,
            id: dto.id,
            status: dto.status,
            success: dto.success,
            toSymbol: dto.toSymbol,
            exchangeApiErrors: dto.exchangeApiErrors,
            maxSpreadPercent: this._decimalDtoConverter.convertFromDto(dto.maxSpreadPercent),
            maxSlippagePercent: this._decimalDtoConverter.convertFromDto(dto.maxSlippagePercent),
            smartRouting: dto.smartRouting,
            triggeredMaxSpread: dto.triggeredMaxSpread,
            triggeredMaxSlippage: dto.triggeredMaxSlippage,
        };
        return result;
    };
    TradeDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            amount: this._decimalDtoConverter.convertToDto(model.amount),
            errorCode: model.errorCode,
            errorMessage: model.errorMessage,
            fromSymbol: model.fromSymbol,
            id: model.id,
            status: model.status,
            success: model.success,
            toSymbol: model.toSymbol,
            exchangeApiErrors: model.exchangeApiErrors,
            maxSpreadPercent: this._decimalDtoConverter.convertToDto(model.maxSpreadPercent),
            maxSlippagePercent: this._decimalDtoConverter.convertToDto(model.maxSlippagePercent),
            smartRouting: model.smartRouting,
            triggeredMaxSpread: model.triggeredMaxSpread,
            triggeredMaxSlippage: model.triggeredMaxSlippage,
        };
        return result;
    };
    return TradeDtoConverter;
}());
exports.TradeDtoConverter = TradeDtoConverter;
