// shrimpy-node/dist/dto-converters/trade-changes-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var balance_change_dto_converter_1 = require("./balance-change-dto-converter");
var trade_dto_converter_1 = require("./trade-dto-converter");
var trade_fill_dto_converter_1 = require("./trade-fill-dto-converter");
var TradeChangesDtoConverter = /** @class */ (function () {
    function TradeChangesDtoConverter() {
        this._balanceChangeDtoConverter = new balance_change_dto_converter_1.BalanceChangeDtoConverter();
        this._tradeDtoConverter = new trade_dto_converter_1.TradeDtoConverter();
        this._tradeFillDtoConverter = new trade_fill_dto_converter_1.TradeFillDtoConverter();
    }
    TradeChangesDtoConverter.prototype.convertFromDto = function (dto) {
        var _this = this;
        var changes = dto.changes.map(function (tradeChangeDto) {
            return _this._balanceChangeDtoConverter.convertFromDto(tradeChangeDto);
        });
        var fills = dto.fills.map(function (tradeFillDto) {
            return _this._tradeFillDtoConverter.convertFromDto(tradeFillDto);
        });
        var result = {
            changes: changes,
            trade: this._tradeDtoConverter.convertFromDto(dto.trade),
            fills: fills,
        };
        return result;
    };
    TradeChangesDtoConverter.prototype.convertToDto = function (model) {
        var _this = this;
        var changes = model.changes.map(function (tradeChange) {
            return _this._balanceChangeDtoConverter.convertToDto(tradeChange);
        });
        var fills = model.fills.map(function (tradeFill) {
            return _this._tradeFillDtoConverter.convertToDto(tradeFill);
        });
        var result = {
            changes: changes,
            trade: this._tradeDtoConverter.convertToDto(model.trade),
            fills: fills,
        };
        return result;
    };
    return TradeChangesDtoConverter;
}());
exports.TradeChangesDtoConverter = TradeChangesDtoConverter;
