// shrimpy-node/dist/dto-converters/user-dto-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var date_dto_converter_1 = require("./date-dto-converter");
var UserDtoConverter = /** @class */ (function () {
    function UserDtoConverter() {
        this._dateDtoConverter = new date_dto_converter_1.DateDtoConverter();
    }
    UserDtoConverter.prototype.convertFromDto = function (dto) {
        var expirationDate;
        if (dto.expirationDate) {
            expirationDate = this._dateDtoConverter.convertFromDto(dto.expirationDate);
        }
        else {
            expirationDate = null;
        }
        var result = {
            expirationDate: expirationDate,
            id: dto.id,
            isEnabled: dto.isEnabled,
            name: dto.name,
        };
        return result;
    };
    UserDtoConverter.prototype.convertToDto = function (model) {
        var result = {
            id: model.id,
            isEnabled: model.isEnabled,
            name: model.name,
        };
        if (model.expirationDate != null) {
            result.expirationDate = this._dateDtoConverter.convertToDto(model.expirationDate);
        }
        return result;
    };
    return UserDtoConverter;
}());
exports.UserDtoConverter = UserDtoConverter;
